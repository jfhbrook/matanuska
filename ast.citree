import { Token } from 'typescript-parsec'
import { TokenKind } from './scanner'

type Expr in './expr' {
  LineNo         => value: number
  DecimalLiteral => value: number
  HexLiteral     => value: number
  OctalLiteral   => value: number
  BinaryLiteral  => value: number
  RealLiteral    => value: number
  BooleanLiteral => value: boolean
  StringLiteral  => value: string
}

type Stmt in './stmt' {
  import { Expr, LineNo } from './expr'

  Line => lineNo: LineNo, command: Expr
  Print => expression: Expr
}
