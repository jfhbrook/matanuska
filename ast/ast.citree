import { Token, TokenKind } from '../tokens'

type Expr in './expr' {
  IntLiteral     => value: number
  RealLiteral    => value: number
  BoolLiteral    => value: boolean
  StringLiteral  => value: string
  NilLiteral!
}

type Cmd in './cmd' {
  import { Expr } from './expr'

  Expression => expression: Expr
  Print      => expression: Expr
}
