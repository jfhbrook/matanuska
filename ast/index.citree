type Expr in './expr' {
  import { TokenKind } from '../tokens'

  Unary          => op: TokenKind, expr: Expr
  Binary         => left: Expr, op: TokenKind, right: Expr
  Logical        => left: Expr, op: TokenKind, right: Expr
  Group          => expr: Expr
  IntLiteral     => value: number
  RealLiteral    => value: number
  BoolLiteral    => value: boolean
  StringLiteral  => value: string
  PromptLiteral  => value: string
  NilLiteral!
}

type Cmd in './cmd' {
  import { Expr } from './expr'

  Cmd        => offsetStart: number = -1, offsetEnd: number = -1
  Expression => expression: Expr
  Print      => expression: Expr
  Exit       => expression: Expr | null
  Rem        => remark: string
}

type Tree in './index' {
  import { Cmd } from './cmd'

  CommandGroup => row: number, source: string, commands: Cmd[]
  Line         => lineNo: number, row: number, source: string, commands: Cmd[]
  Input        => input: Array<CommandGroup | Line>
  Program      => lines: Line[]
}
