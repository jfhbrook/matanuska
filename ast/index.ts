import { Token, TokenKind } from '../tokens';
import { Cmd } from './cmd';

export interface TreeVisitor<R> {
  visitCommandGroupTree(node: CommandGroup): R;
  visitLineTree(node: Line): R;
  visitProgramTree(node: Program): R;
}

export abstract class Tree {
  abstract accept<R>(visitor: TreeVisitor<R>): R;
}

export class CommandGroup extends Tree {
  constructor(public readonly commands: Cmd[]) {
    super();
  }

  accept<R>(visitor: TreeVisitor<R>): R {
    return visitor.visitCommandGroupTree(this);
  }
}

export class Line extends Tree {
  constructor(
    public readonly lineNo: number,
    public readonly commands: Cmd[],
  ) {
    super();
  }

  accept<R>(visitor: TreeVisitor<R>): R {
    return visitor.visitLineTree(this);
  }
}

export class Program extends Tree {
  constructor(public readonly lines: Line[]) {
    super();
  }

  accept<R>(visitor: TreeVisitor<R>): R {
    return visitor.visitProgramTree(this);
  }
}
