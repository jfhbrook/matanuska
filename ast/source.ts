//
// A tuple-like class to store source code. This is needed so we can easily
// renumber source lines without having to use regular expressions.
//

export class Source {
  constructor(
    public leadingWs: string,
    public lineNo: string,
    public separatingWs: string,
    public source: string,
  ) {}

  public clone(): Source {
    return new Source(
      this.leadingWs,
      this.lineNo,
      this.separatingWs,
      this.source,
    );
  }

  public toString(): string {
    return [this.leadingWs, this.lineNo, this.separatingWs, this.source].join(
      '',
    );
  }
}
