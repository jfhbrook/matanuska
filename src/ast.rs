// This contains stuff the parser can return.
//
// This file needs to be fleshed out in order to write the parser against it,
// but the parser itself is a very bad reference for most of this file. Chances
// are that the actual data structures exist in other files in yabasic; as I
// find them, I can write them down here.

use thiserror::Error;

use crate::tokens::Ident;

// As mentioned in crate::error, exception handling needs a lot of work.
//
// Right now, all this can really do is use thiserror to represent a few
// simple error types which can be represented in the language. But really
// I need something more like Python's errors - a struct with a code, a
// message and a traceback. That means awareness of the line which raised the
// error.
//
// It also needs to represent all errors possible, not just ones that are
// easy to clone. That's in part because I don't want two classes of errors,
// but also because I want those errors representable in the language.
//
// The punchline is likely a bunch of "manual" conversion code for various
// kinds of errors.
//
// A final note: in s7bas, I used exceptions for non-error cases like
// Python does. That includes keyboard interrupts of course (consistent with
// rustyline) but also execution cases like Halt and parse cases like Eof.
// I'm not sure I want to replicate that here.
//
// Note that miette might be useful here.
#[derive(Error, PartialEq, Debug, Clone)]
pub(crate) enum Exception {
    #[error("parse error")]
    ParseError(String),

    #[error("value error")]
    ValueError(String),

    #[error("name error")]
    NameError(String, String),

    #[error("file error")]
    FileError(String),

    // These can be triggered by errors from rustyline
    #[error("keyboard interrupt")]
    KeyboardInterrupt,

    #[error("EOF")]
    Eof,
    // #[error("halt")]
    // Halt(Interpreter<GlobalContext>),

    // #[error("goto")]
    // Goto(i64, Interpreter<GlobalContext>),
}

// Literal values. This is somewhat borrowed from yabasic, but will probably
// end up evolving a lot.
#[derive(PartialEq, Debug, Clone)]
pub(crate) enum Literal {
    Const(f64),
    Float(f64),
    Int(i64),
    String(String),
}

// yabasic calls these identifiers. s7bas didn't really have this concept, and
// in general there's some semantic confusion between symbols, variables,
// identifiers and values.
//
// my read is:
//
// - variables are specific to things you can assign and read from as
//   variables in the language
// - symbols are anything that has an identifier (called a "name" in the
//   current lexer
// - identifiers are "names" for symbols
// - values are what symbols actually contain
//
// TODO: iron out these semantic issues, here and in the lexer.
#[derive(PartialEq, Debug, Clone)]
pub(crate) enum Identifier {
    SymbolOrLineno(Ident),
    FunctionName(Ident),
    FunctionOrArray(Ident),
    StringFunctionOrArray(Ident),
}

// In a classic BASIC, the basic building blocks are called instructions.
#[derive(Debug, Clone)]
pub enum Instruction {
    Print(Option<i64>, Vec<Expr>),
    Remark(String),
    // Let(Id, Expr),

    // Conditionals (see: Boolean logic, comparisons)
    // If(Expr, Box<Instruction>),

    // Goto(Expr),

    // For(Id, Expr, Expr, Option<i64>),
    // Next(Vec<Id>),
    // End,

    // Gosub(Expr),
    // Return,

    // Dim(Id, Vec<i64>),
    // Data(Vec<Constant>),

    // Input(Option<i64>, Vec<Id>),
    // Open(Value, Access, FileDescriptor),
    // Close(FileDescriptor),
    // Poke(Vec<Value>),
    // Read(Vec<Id>),
    List(Option<i64>, Option<i64>),
    Save(String),
    Load(String),
    Run(Option<i64>),
    // Restore,
    // Stop,

    // Sys(Value),
    // Wait(Vec<Value>),
    Expr(Expr),
    // TODO: yabasic supports "docstrings". Kinda a cool feature - do I
    // want it?
    // Docu(String)
}

// Commands are made up of a series of Instructions separated by colon
// separators. These are generally numbered and entered as Lines.
#[derive(Debug, Clone)]
pub struct Command {
    instructions: Vec<Instruction>,
}

impl Command {
    fn new(instructions: Vec<Instruction>) -> Command {
        Command { instructions }
    }
}

// Instructions may contain expressions. Expressions can be evaluated into
// values. These are all expression types I intended to define in s7bas.
#[derive(Debug, Clone)]
pub enum Expr {
    // Operators, in order of precedence in Visual Basic. Operators of equal
    // precedence are evaluated left to right.

    // 0. Value literals
    Literal(Literal),
    Ident(Ident),

    // 1. parens
    Parens(Box<Expr>),

    // 2. VB has an Await keyword...
    // Await(Box<Expr>)
    // (other keywords here too, such as...)
    // Call(Vec<Expr>, IndexMap<String, Expr>)

    // 3. Exponentiation
    Pow(Box<Expr>, Box<Expr>),

    // 4. +/- signs
    Negative(Box<Expr>),
    Positive(Box<Expr>),

    // 5. times/divide
    Multiply(Box<Expr>, Box<Expr>),
    Divide(Box<Expr>, Box<Expr>),
    // (A .\ b as solving for x in Ax=b, as in MATLAB)
    LeftDivide(Box<Expr>, Box<Expr>),
    // (A .* B) as element-wise multiplication, as in MATLAB)
    ElementWiseMultiply(Box<Expr>, Box<Expr>),

    // 6. VB has a Mod operator
    Mod(Box<Expr>, Box<Expr>),

    // 7. add/subtract
    Add(Box<Expr>, Box<Expr>),
    Subtract(Box<Expr>, Box<Expr>),
    // 8. string concatenation - VB has a separate cat operator, however
    // OG basic dialects overload + - but this is always an option!
    // Cat(Box<Expr>, Box<Expr>)

    // 9. bitshift operators
    LeftShift(Box<Expr>, Box<Expr>),
    RightShift(Box<Expr>, Box<Expr>),

    // 10. Comparisons
    Eq(Box<Expr>, Box<Expr>),
    Ne(Box<Expr>, Box<Expr>),
    Lt(Box<Expr>, Box<Expr>),
    Gt(Box<Expr>, Box<Expr>),
    Le(Box<Expr>, Box<Expr>),
    Ge(Box<Expr>, Box<Expr>),
    // VB does Is/IsNot for object identity, Like, etc

    // 11. Logical operators
    And(Box<Expr>, Box<Expr>),
    Or(Box<Expr>, Box<Expr>),
    Xor(Box<Expr>, Box<Expr>),
    Not(Box<Expr>),
}

// Lines are "numbered commands".
#[derive(Debug, Clone)]
pub struct Line {
    line_no: u16,
    command: Command,
}
