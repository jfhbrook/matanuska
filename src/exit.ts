import { ExitCode, ExitCoded } from './sysexits';

/**
 * When thrown, causes the program to exit successfully. Takes an optional
 * message.
 */
export class Exit extends Error implements ExitCoded {
  public readonly exitCode = ExitCode.Success;

  constructor(message?: string) {
    super(message || '');
  }
}
