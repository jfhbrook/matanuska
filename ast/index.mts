import { Source } from './source.mjs';
import { Instr } from './instr.mjs';

export interface TreeVisitor<R> {
  visitCmdTree(node: Cmd): R;
  visitLineTree(node: Line): R;
  visitInputTree(node: Input): R;
  visitProgramTree(node: Program): R;
}

export abstract class Tree {
  abstract accept<R>(visitor: TreeVisitor<R>): R;
}

export class Cmd extends Tree {
  constructor(
    public cmdNo: number,
    public row: number,
    public source: Source,
    public instructions: Instr[],
  ) {
    super();
  }

  accept<R>(visitor: TreeVisitor<R>): R {
    return visitor.visitCmdTree(this);
  }
}

export class Line extends Tree {
  constructor(
    public lineNo: number,
    public row: number,
    public source: Source,
    public instructions: Instr[],
  ) {
    super();
  }

  accept<R>(visitor: TreeVisitor<R>): R {
    return visitor.visitLineTree(this);
  }
}

export class Input extends Tree {
  constructor(public input: Array<Cmd | Line>) {
    super();
  }

  accept<R>(visitor: TreeVisitor<R>): R {
    return visitor.visitInputTree(this);
  }
}

export class Program extends Tree {
  constructor(
    public filename: string,
    public lines: Line[],
  ) {
    super();
  }

  accept<R>(visitor: TreeVisitor<R>): R {
    return visitor.visitProgramTree(this);
  }
}
