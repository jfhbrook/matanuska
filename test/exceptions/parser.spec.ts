import { describe, test } from 'vitest';
import { t } from '../helpers/tap';

import { ExitCode } from '../../exit';

import {
  SyntaxError,
  ParseError,
  SyntaxWarning,
  ParseWarning,
  mergeParseErrors,
  splitParseError,
  removeFromParseError,
} from '../../exceptions';
import { Source } from '../../ast/source';
import { FILENAME } from '../helpers/files';

describe('ParseError', () => {
  test('it can construct a ParseError', () => {
    const line = new Source('', '100', ' ', 'print someFn(ident');

    const exc = new ParseError([
      new SyntaxError('expected )', {
        filename: FILENAME,
        row: 0,
        isLine: false,
        lineNo: 100,
        cmdNo: null,
        offsetStart: 22,
        offsetEnd: 23,
        source: line,
      }),
      new SyntaxWarning('identifier has no sigil', {
        filename: FILENAME,
        row: 0,
        isLine: false,
        lineNo: 100,
        cmdNo: null,
        offsetStart: 17,
        offsetEnd: 18,
        source: line,
      }),
    ]);

    t.ok(exc);
    t.equal(exc.message, '');
    t.equal(exc.exitCode, ExitCode.Software);
    t.equal(exc.traceback, null);
    t.equal(exc.errors.length, 2);

    t.equal(exc.errors[0].message, 'expected )');
    t.same(exc.errors[0].filename, FILENAME);
    t.equal(exc.errors[0].lineNo, 100);
    t.same(exc.errors[0].offsetStart, 22);
    t.same(exc.errors[0].offsetEnd, 23);

    t.equal(exc.errors[1].message, 'identifier has no sigil');
    t.same(exc.errors[1].filename, FILENAME);
    t.equal(exc.errors[1].lineNo, 100);
    t.same(exc.errors[1].offsetStart, 17);
    t.same(exc.errors[1].offsetEnd, 18);
  });
});

describe('ParseWarning', () => {
  test('it can construct a ParseWarning', () => {
    const line = new Source('', '100', ' ', 'print someFn(ident)');

    const exc = new ParseWarning([
      new SyntaxWarning('identifier has no sigil', {
        filename: FILENAME,
        row: 0,
        isLine: false,
        lineNo: 100,
        cmdNo: null,
        offsetStart: 17,
        offsetEnd: 18,
        source: line,
      }),
    ]);

    t.ok(exc);
    t.equal(exc.message, '');
    t.equal(exc.traceback, null);
    t.equal(exc.warnings.length, 1);

    t.equal(exc.warnings[0].message, 'identifier has no sigil');
    t.same(exc.warnings[0].filename, FILENAME);
    t.equal(exc.warnings[0].lineNo, 100);
    t.same(exc.warnings[0].offsetStart, 17);
    t.same(exc.warnings[0].offsetEnd, 18);
  });
});

test('mergeParseErrors', () => {
  const PARSE_WARNING_1 = new ParseWarning([
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 0,
      isLine: true,
      lineNo: 100,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '100', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 4,
      isLine: true,
      lineNo: 500,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '500', ' ', 'print someFn(ident)'),
    }),
  ]);

  const PARSE_WARNING_2 = new ParseWarning([
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 1,
      isLine: true,
      lineNo: 200,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '200', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 5,
      isLine: true,
      lineNo: 600,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '600', ' ', 'print someFn(ident)'),
    }),
  ]);

  const PARSE_ERROR_1 = new ParseError([
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 2,
      isLine: false,
      lineNo: 300,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '300', ' ', 'print someFn(ident'),
    }),
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 6,
      isLine: false,
      lineNo: 700,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '700', ' ', 'print someFn(ident'),
    }),
  ]);

  const PARSE_ERROR_2 = new ParseError([
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 3,
      isLine: false,
      lineNo: 400,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '400', ' ', 'print someFn(ident'),
    }),
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 7,
      isLine: false,
      lineNo: 800,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '800', ' ', 'print someFn(ident'),
    }),
  ]);

  t.matchSnapshot(
    mergeParseErrors([PARSE_WARNING_2, PARSE_WARNING_1]),
    'merge two warnings',
  );
  t.matchSnapshot(
    mergeParseErrors([PARSE_ERROR_2, PARSE_ERROR_1]),
    'merge two errors',
  );
  t.matchSnapshot(
    mergeParseErrors([PARSE_WARNING_1, PARSE_ERROR_1]),
    'merge a warning and an error',
  );
  t.matchSnapshot(
    mergeParseErrors([PARSE_WARNING_1, null]),
    'merge a warning and null',
  );
  t.matchSnapshot(
    mergeParseErrors([PARSE_WARNING_1, null]),
    'merge an error and null',
  );
  t.matchSnapshot(
    mergeParseErrors([PARSE_ERROR_1, PARSE_WARNING_2, null]),
    'merge a warning, an error and null',
  );
  t.matchSnapshot(mergeParseErrors([null, null, null]), 'merge a few nulls');
});

test('splitParseError', () => {
  const ERROR = new ParseError([
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 0,
      isLine: true,
      lineNo: 100,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '100', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 1,
      isLine: true,
      lineNo: 200,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '200', ' ', 'print someFn(ident)'),
    }),

    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 2,
      isLine: false,
      lineNo: 300,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '300', ' ', 'print someFn(ident'),
    }),
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 3,
      isLine: false,
      lineNo: 400,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '400', ' ', 'print someFn(ident'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 4,
      isLine: true,
      lineNo: 500,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '500', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 5,
      isLine: true,
      lineNo: 600,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '600', ' ', 'print someFn(ident)'),
    }),
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 6,
      isLine: false,
      lineNo: 700,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '700', ' ', 'print someFn(ident'),
    }),
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 7,
      isLine: false,
      lineNo: 800,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '800', ' ', 'print someFn(ident'),
    }),
  ]);

  const WARN = new ParseError([
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 0,
      isLine: true,
      lineNo: 100,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '100', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 1,
      isLine: true,
      lineNo: 200,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '200', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 4,
      isLine: true,
      lineNo: 500,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '500', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 5,
      isLine: true,
      lineNo: 600,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '600', ' ', 'print someFn(ident)'),
    }),
  ]);

  function testSplit(error: Record<number, any>, key: string) {
    for (const [k, err] of Object.entries(error)) {
      const errs = err.errors ? err.errors : err.warnings;
      for (const e of errs) {
        t.equal(String(k), String(e[key]));
      }
    }
  }

  testSplit(splitParseError(ERROR, 'row'), 'row');
  testSplit(splitParseError(WARN, 'row'), 'row');
  testSplit(splitParseError(ERROR, 'lineNo'), 'lineNo');
  testSplit(splitParseError(WARN, 'lineNo'), 'lineNo');
  t.same(splitParseError(null, 'row'), {});
});

test('removeFromParseError', () => {
  const ERROR = new ParseError([
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 0,
      isLine: true,
      lineNo: 100,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '100', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 1,
      isLine: true,
      lineNo: 200,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '200', ' ', 'print someFn(ident)'),
    }),

    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 2,
      isLine: false,
      lineNo: 300,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '300', ' ', 'print someFn(ident'),
    }),
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 3,
      isLine: false,
      lineNo: 400,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '400', ' ', 'print someFn(ident'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 4,
      isLine: true,
      lineNo: 500,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '500', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 5,
      isLine: true,
      lineNo: 600,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '600', ' ', 'print someFn(ident)'),
    }),
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 6,
      isLine: false,
      lineNo: 700,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '700', ' ', 'print someFn(ident'),
    }),
    new SyntaxError('expected )', {
      filename: FILENAME,
      row: 7,
      isLine: false,
      lineNo: 800,
      cmdNo: null,
      offsetStart: 22,
      offsetEnd: 23,
      source: new Source('', '800', ' ', 'print someFn(ident)'),
    }),
  ]);

  const WARN = new ParseError([
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 0,
      isLine: true,
      lineNo: 100,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '100', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 1,
      isLine: true,
      lineNo: 200,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '200', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 4,
      isLine: true,
      lineNo: 500,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '500', ' ', 'print someFn(ident)'),
    }),
    new SyntaxWarning('identifier has no sigil', {
      filename: FILENAME,
      row: 5,
      isLine: true,
      lineNo: 600,
      cmdNo: null,
      offsetStart: 17,
      offsetEnd: 18,
      source: new Source('', '600', ' ', 'print someFn(ident)'),
    }),
  ]);

  function testRemove(error: any, key: string, value: number) {
    const errs = error.errors ? error.errors : error.warnings;
    for (const e of errs) {
      t.notEqual(String(e[key]), String(value));
    }
  }

  testRemove(removeFromParseError(ERROR, 'row', 2), 'row', 2);
  testRemove(removeFromParseError(WARN, 'row', 2), 'row', 2);
  t.same(removeFromParseError(null, 'row', 2), null);
  testRemove(removeFromParseError(ERROR, 'lineNo', 400), 'lineNo', 400);
  testRemove(removeFromParseError(WARN, 'lineNo', 400), 'lineNo', 400);
  t.same(removeFromParseError(null, 'lineNo', 400), null);
});
