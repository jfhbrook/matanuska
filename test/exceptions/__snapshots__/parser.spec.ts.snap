// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mergeParseErrors 1`] = `[ParseWarning]`;

exports[`mergeParseErrors 2`] = `[Exception]`;

exports[`mergeParseErrors 3`] = `[Exception]`;

exports[`mergeParseErrors 4`] = `[ParseWarning]`;

exports[`mergeParseErrors 5`] = `[ParseWarning]`;

exports[`mergeParseErrors 6`] = `[Exception]`;

exports[`mergeParseErrors 7`] = `null`;
